File: 1501~2000/1727. Largest Submatrix With Rearrangements/main.py
"""
Source: https://leetcode.com/problems/largest-submatrix-with-rearrangements/
Date: 2023/4/15
Skill:
Ref:
Runtime: 1284 ms, faster than 88.24%
Memory Usage: 37.5 MB, less than 88.24%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def largestSubmatrix(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        hist, res = [0 for _ in range(n)], 0
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 1:
                    hist[j] += 1
                else:
                    hist[j] = 0
            cur_row = hist.copy()
            cur_row.sort(reverse=True)
            for k in range(n):
                res = max(res, cur_row[k] * (k + 1))

        return res


if __name__ == "__main__":
    s = Solution()

File: 1501~2000/1727. Largest Submatrix With Rearrangements/main.cpp
/**
 * Source: jpeg.ly/r9_ca
 * Date: 2023/11/26
 * Skill: Histogram
 * Ref:
 * Runtime: 150 ms, faster than 77.84% of C++ online submissions
 * Memory Usage: 66.32 MB, less than 70.45% of C++ online submissions
 * Time complexity:
 * Space complexity:
 * Constraints:
 *
 */


#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;
#define SZ(X) ((int)(X).size())


class Solution {
public:
    int largestSubmatrix(vector<vector<int>>& matrix) {
        int m = SZ(matrix), n = SZ(matrix[0]), res = 0;
        for (int i = 1; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] > 0) {
                    matrix[i][j] += matrix[i - 1][j];
                }
            }
        }
        for (int i = m - 1; i >= 0; --i) {
            sort(begin(matrix[i]), end(matrix[i]));
            for (int j = 0; j < n; ++j) {
                res = max(res, (n - j) * matrix[i][j]);
            }
        }
        return res;
    }
};